system/lib/libc/musl/src/thread/pthread_getspecific.c

```diff
@@ -4,6 +4,10 @@
 static void *__pthread_getspecific(pthread_key_t k)
 {
 	struct pthread *self = __pthread_self();
+	// XXX EMSCRIPTEN: self->tsd can be NULL in the case of the
+	// main thread where pthread_key_create is not called.
+	// See __pthread_key_create for where it get assigned.
+	if (!self->tsd) return NULL;
```

Comment:
Instead of this check, perhaps we could ensure that `self->tsd` is always assigned? This also matches musl's convention, which seems to always assign it to `__pthread_tsd_main` once the first thread is created:
https://github.com/emscripten-core/emscripten/blob/4d864df0a57024d667e8db171aab2c3385d04c30/system/lib/libc/musl/src/thread/pthread_create.c#L162-L163
https://github.com/emscripten-core/emscripten/blob/4d864df0a57024d667e8db171aab2c3385d04c30/system/lib/libc/musl/src/thread/pthread_create.c#L201

Reply:
The problem is that we have at least one test that calls `pthread_getspecific` before any threads are created and expects it to work: posixtest.test_pthread_key_create_2_1

Reply:
Or at least the test expects it to return NULL and not crash

Reply:
Ah I see, that test looks odd. When compiling with GCC it also complains that key is not initialized.
```bash
conformance/interfaces/pthread_key_create/2-1.c:33:14: error: ‘key’ is used uninitialized [-Werror=uninitialized]
   33 |         rc = pthread_getspecific(key);
      |              ^~~~~~~~~~~~~~~~~~~~~~~~
conformance/interfaces/pthread_key_create/2-1.c:29:23: note: ‘key’ declared here
   29 |         pthread_key_t key;
      |                       ^~~
cc1: all warnings being treated as errors
``` 

But running it locally and on Alpine (which also uses musl) can be done without any problems.
```bash
$ cd ./tests/third_party/posixtestsuite
$ docker run -v $(pwd):/app -w /app -it alpine:latest \
      apk add build-base && \
      POSIX_TARGET=conformance/interfaces/pthread_key_create make LDFLAGS='-pthread -Wno-error=uninitialized'
...
conformance/interfaces/pthread_key_create/2-1: build: PASS
conformance/interfaces/pthread_key_create/2-1: link: PASS
...
conformance/interfaces/pthread_key_create/2-1: execution: PASS
```

Do you think it makes sense to initialize `tsd` in `__emscripten_pthread_data_constructor`? Similar to the logic in `pthread_key_create`.
https://github.com/emscripten-core/emscripten/blob/8192916e32e3b74edadb147e545804b52d652d7a/system/lib/libc/musl/src/thread/pthread_key_create.c#L22-L24

Though, the current approach would also be sufficient.

Reply:
Oh, my test command for Alpine was wrong (it ran the test on my own system). This also seems to segfault on Alpine.
```bash
$ docker run -v $(pwd):/app -w /app -it alpine:latest sh -c "\
    apk add build-base && \
    POSIX_TARGET=conformance/interfaces/pthread_key_create make LDFLAGS='-pthread -Wno-error=uninitialized'"
$ cat logfile
conformance/interfaces/pthread_key_create/1-1: build: PASS
conformance/interfaces/pthread_key_create/1-1: link: PASS
conformance/interfaces/pthread_key_create/1-2: build: PASS
conformance/interfaces/pthread_key_create/1-2: link: PASS
conformance/interfaces/pthread_key_create/2-1: build: PASS
conformance/interfaces/pthread_key_create/2-1: link: PASS
conformance/interfaces/pthread_key_create/3-1: build: PASS
conformance/interfaces/pthread_key_create/3-1: link: PASS
conformance/interfaces/pthread_key_create/speculative/5-1: build: PASS
conformance/interfaces/pthread_key_create/speculative/5-1: link: PASS
conformance/interfaces/pthread_key_create/1-1: execution: PASS
conformance/interfaces/pthread_key_create/1-2: execution: PASS
conformance/interfaces/pthread_key_create/2-1: execution: INTERRUPTED: Output: 
Segmentation fault (core dumped)
conformance/interfaces/pthread_key_create/3-1: execution: PASS
conformance/interfaces/pthread_key_create/speculative/5-1: execution: FAILED: Output: 
Test FAILED: EAGAIN was returned before the key limit was exceeded
```

Perhaps we should report this upstream.